compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -645,3 +657,93 @@ fn may_define_opaque_type(tcx: TyCtxt<'_>, def_id: LocalDefId, opaque_hir_id: hi
     );
     res
 }
+
+/// Checks that the item may register hidden types for the
+/// opaque type, if the opaque type shows up in its signature.
+#[instrument(level = "debug", skip(tcx), ret)]
+pub fn may_define_impl_trait_in_assoc_ty_modulo_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    opaque_def_id: LocalDefId,
+) -> Option<impl TypeVisitable<TyCtxt<'tcx>>> {
+    let sig = match tcx.def_kind(def_id) {
+        DefKind::AssocFn => Ok(tcx.fn_sig(def_id).skip_binder()),
+        DefKind::AssocConst | DefKind::AssocTy => Err(tcx.type_of(def_id).skip_binder()),
```

Comment:
why `DefKid::AssocTy`?

Reply:
I added a comment for that. The reason for that is that we do wf checks of opaque types within their parent. So we need to treat their parent as having them in the signature, as we'd otherwise fail that wf check. The wfcheck is fishy anyway and we're working on removing it, so that will go away once we remove the wfcheck